### PR TITLE
jenkins will build python.

### PR DESCRIPTION
**Pull request by stefoss23**
Adding python testing to jenkins build system.

### CODE REVIEW

File: jenkins/build-opm-module.sh [hunk 1]
```diff
@@ -4,6 +4,7 @@ declare -A configurations
 
 declare -A EXTRA_MODULE_FLAGS
 EXTRA_MODULE_FLAGS[opm-simulators]="-DBUILD_EBOS_EXTENSIONS=ON -DBUILD_EBOS_DEBUG_EXTENSIONS=ON"
+EXTRA_MODULE_FLAGS[opm-common] = "-DENABLE_PYTHON=ON"
```
Comment (akva2) [inline jenkins/build-opm-module.sh]: no space around assignment - bash will not eat it.

**Comment by joakim-hove** [comment]
When/if this is merged jenkins will build and test opm-common with Python. The default switch will still be `ENABLE_PYTHON=OFF` - i.e. unless you explicitly ask for it there will be no snakes in your office.

**Comment by stefoss23** [comment]
jenkins build this please

File: jenkins/build-opm-module.sh [hunk 2]
```diff
@@ -4,7 +4,7 @@ declare -A configurations
 
 declare -A EXTRA_MODULE_FLAGS
 EXTRA_MODULE_FLAGS[opm-simulators]="-DBUILD_EBOS_EXTENSIONS=ON -DBUILD_EBOS_DEBUG_EXTENSIONS=ON"
-EXTRA_MODULE_FLAGS[opm-common] = "-DENABLE_PYTHON=ON"
+EXTRA_MODULE_FLAGS[opm-common]="-DENABLE_PYTHON=ON"
```
Comment (akva2) [inline jenkins/build-opm-module.sh]: EXTRA_MODULE_FLAGS[libecl]="-DCMAKE_POSITION_INDEPENDENT_CODE=1"

**Comment by stefoss23** [comment]
jenkins build this please

**Comment by stefoss23** [comment]
stop

**Comment by akva2** [comment]
how do you plan to handle the python version mess? it seems pybind11 used from source tree has no such support. i'd like to see it spit out py2 and py3 without having to rebuild (for packaging reasons)..

**Comment by joakim-hove** [comment]
> how do you plan to handle the python version mess? it seems pybind11 used from source tree has no such support. 

This is not extensively thought through; my (limited - but still) experience with pybind11 has been that py2 and py3 interchange quite well - you will essentially build against whatever Python libs and header is found on your system - and my experience has been that things Just Work&trade; from there

> i'd like to see it spit out py2 and py3 without having to rebuild (for packaging reasons)..

Now - that is a higher order problem which has not been in our head yet; would be good if we could get the basic stuff working first, and then subsequently address issues like this?

**Comment by akva2** [comment]
yes, for sure. i'm just worried the current solution is fundamentally unable to do so. the problem is that pybind11 records the available version when you add subdirectory. i've tried to do the unset mess they point to on the pybind11 github issue tracker https://github.com/pybind/pybind11/issues/748, but it gets rather meh rather quick due to the amount of targets that needs duplication.

alternatively, we ignore py2 and only go py3. i'm fine with that solution as such, but py2 only (which is the default on anything but fedora still) is not acceptable imo.

**Comment by atgeirr** [comment]
Not sure if this is the place to ask, but do you understand the link warning I get from the brand new sunbeam stuff?

```
ld: warning: direct access in function 'boost::system::error_category::std_category::equivalent(int, std::__1::error_condition const&) const' from file 'CMakeFiles/libsunbeam.dir/eclipse_state.cpp.o' to global weak symbol 'typeinfo for boost::system::error_category::std_category' from file '../../lib/libopmcommon.a(EclipseState.cpp.o)' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in function 'boost::system::error_category::std_category::equivalent(std::__1::error_code const&, int) const' from file 'CMakeFiles/libsunbeam.dir/eclipse_state.cpp.o' to global weak symbol 'typeinfo for boost::system::error_category::std_category' from file '../../lib/libopmcommon.a(EclipseState.cpp.o)' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in function 'unsigned long std::__1::__str_rfind<char, unsigned long, std::__1::char_traits<char>, 18446744073709551615ul>(char const*, unsigned long, char const*, unsigned long, unsigned long)' from file 'CMakeFiles/libsunbeam.dir/eclipse_state.cpp.o' to global weak symbol 'std::__1::char_traits<char>::eq(char, char)' from file '../../lib/libopmcommon.a(ParseContext.cpp.o)' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
...
```
And plenty more like it. I also get compiler warnings (probably clang-only if gcc did not see it):
```
/Users/atgeirr/alt-opm/src/opm-common/python/pybind11/include/pybind11/pybind11.h:1727:50: warning: modifying constructor parameter 'flag' that
      shadows a field of 'set_flag' [-Wshadow-field-in-constructor-modified]
        set_flag(bool &flag) : flag(flag) { flag = true; }
                                                 ^
/Users/atgeirr/alt-opm/src/opm-common/python/pybind11/include/pybind11/pybind11.h:1727:24: note: variable 'flag' is declared here
        set_flag(bool &flag) : flag(flag) { flag = true; }
                       ^
/Users/atgeirr/alt-opm/src/opm-common/python/pybind11/include/pybind11/pybind11.h:1726:15: note: previous declaration is here
        bool &flag;
              ^
```

**Comment by joakim-hove** [comment]
> Not sure if this is the place to ask, but do you understand the link warning I get from the brand new sunbeam stuff?

This should not have reached @atgeirr at all - fix coming. Actually ignoring the warnings will be a next step.

**Comment by joakim-hove** [comment]
Please see: #867

**Comment by akva2** [comment]
python-decorator has been installed on jenkins.

as for the warnings, they are caused by the introduction of -Wshadow by default. suggested solution:

```diff
diff --git a/python/sunbeam/CMakeLists.txt b/python/sunbeam/CMakeLists.txt
index 3d519f95..14e48f18 100644
--- a/python/sunbeam/CMakeLists.txt
+++ b/python/sunbeam/CMakeLists.txt
@@ -16,6 +16,9 @@ pybind11_add_module(libsunbeam sunbeam.cpp
 
 set_target_properties( libsunbeam PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/python/sunbeam )
 
+# Removed due to a lot of shadowing warnings in pybind11
+string(REPLACE "-Wshadow" "" FLAGS_REPLACED ${CMAKE_CXX_FLAGS})
+set_target_properties(libsunbeam PROPERTIES COMPILE_FLAGS ${FLAGS_REPLACED})
+
 # The cmake config files generated by opm-common are generated in a non standard
 # way and they are not fully relocatable. This creates problems when we are
 # building sunbeam towards a opm-common distribution which has been installed
```

**Comment by joakim-hove** [comment]
> i'm just worried the current solution is fundamentally unable to do so

Have discussed a `setup.py` based solution on :phone: - seems to be a viable long term solution.

**Comment by stefoss23** [comment]
jenkins build this please

**Comment by akva2** [comment]
no -V wanted, nor needed. all test results including the terminal outputs are readily available on jenkins.

**Comment by joakim-hove** [comment]
The - V was just temporary - poi t is it takes quite long before the Test Results become available.

**Comment by akva2** [comment]
sure, just making sure you don't slip it in cause it was not commented on :)

**Comment by joakim-hove** [comment]
I don't really understand what you mean with Python decorator in this context?

**Comment by akva2** [comment]
the test scripts has 'import decorator' in them - this is not part of the python stdlib - so i had to apt install python-decorator.

**Comment by joakim-hove** [comment]
> the test scripts has 'import decorator' in them

ahhh - sorry for being dense; those things should - at the very least - be more explicit when the `setup.py` solution is in place.

**Comment by stefoss23** [comment]
jenkins build this please

**Comment by stefoss23** [comment]
jenkins build this please

**Comment by stefoss23** [comment]
jenkins build this please

**Comment by akva2** [comment]
please include the -Wshadow suppressing posted above in some form or another.

**Comment by stefoss23** [comment]
jenkins build this please

**Pull request merged**